packages/e2e-test-utils/src/rest-api.js

```diff
@@ -0,0 +1,116 @@
+/**
+ * External dependencies
+ */
+import fetch from 'node-fetch';
+import FormData from 'form-data';
+
+/**
+ * WordPress dependencies
+ */
+import apiFetch from '@wordpress/api-fetch';
+
+/**
+ * Internal dependencies
+ */
+import { WP_BASE_URL } from './shared/config';
+import { createURL } from './create-url';
+
+// `apiFetch` expects `window.fetch` to be available in its default handler.
+global.window = global.window || {};
+global.window.fetch = fetch;
+
+const setAPIRootURL = ( async () => {
+	// Discover the API root url using link header.
+	// See https://developer.wordpress.org/rest-api/using-the-rest-api/discovery/#link-header
+	const res = await fetch( WP_BASE_URL, { method: 'HEAD' } );
+	const links = res.headers.get( 'link' );
+	const restLink = links.match( /<([^>]+)>; rel="https:\/\/api\.w\.org\/"/ );
+
+	if ( ! restLink ) {
+		throw new Error( `Failed to discover REST API endpoint.
+Link header: ${ links }` );
+	}
+
+	const [ , rootURL ] = restLink;
+	apiFetch.use( apiFetch.createRootURLMiddleware( rootURL ) );
+} )();
+
+const setNonce = ( async () => {
+	const formData = new FormData();
+	formData.append( 'log', 'admin' );
+	formData.append( 'pwd', 'password' );
+
+	// Login to admin using fetch.
+	const loginResponse = await fetch( createURL( 'wp-login.php' ), {
+		method: 'POST',
+		headers: formData.getHeaders(),
+		body: formData,
+		redirect: 'manual',
+	} );
+
+	// Retrieve the cookies.
+	const cookies = loginResponse.headers.get( 'set-cookie' );
+	const cookie = cookies
+		.split( ',' )
+		.map( ( setCookie ) => setCookie.split( ';' )[ 0 ] )
+		.join( ';' );
+
+	apiFetch.nonceEndpoint = createURL(
+		'wp-admin/admin-ajax.php',
+		'action=rest-nonce'
+	);
+
+	// Get the initial nonce.
+	const res = await fetch( apiFetch.nonceEndpoint, {
+		headers: { cookie },
+	} );
+	const nonce = await res.text();
+
+	// Register the nonce middleware.
+	apiFetch.use( apiFetch.createNonceMiddleware( nonce ) );
+
+	// For the nonce to work we have to also pass the cookies.
+	apiFetch.use( function setCookieMiddleware( request, next ) {
+		return next( {
+			...request,
+			headers: {
+				...request.headers,
+				cookie,
+			},
+		} );
+	} );
+} )();
+
+/**
+ * Call REST API using `apiFetch` to build and clear test states.
+ *
+ * @param {Object} options `apiFetch` options.
+ * @return {Promise<any>} The response value.
+ */
+async function rest( options = {} ) {
+	// Only need to set them once but before any requests.
+	await Promise.all( [ setAPIRootURL, setNonce ] );
+
+	return await apiFetch( options );
```

Comment:
Would it be possible to use `wp.apiFetch` instead that should have already all params set when the user is logged in:
```js
await page.evaluate( ( params ) =>
	wp.apiFetch( params ),
	options
);
 ```

Reply:
That would make the test depends on the page it's currently running on. Some pages don't have `wp.apiFetch` available by default, and sometimes the logged in user doesn't have the necessary permissions. It also means that a reload is required after making such requests. Calling it server-side seems like a more robust approach to me.

Reply:
I must admit that it's all confusing. In the current shape, the `rest` method is doing a REST API request as an admin with hardcoded credentials to `admin` + `password`. If the goal is to make it general purpose then it's far from it. Some concerns:
- e2e tests can be executed also with sites that don't have the admin account with the default credentials
- should it be possible to use `rest` with a different user account or as logged out?
- `wp-login.php` set the cookie so it's very likely that once the test refreshes the page then the rest of the test is going to be performed as an admin

Reply:
`rest` isn't the best name here. It's running in the nodejs process rather than the browser context, so it's independent to the test and won't affect the current logged in user. As mentioned in the description of the PR, it should only be used to setup/clear states between or during tests. I'll keep searching for a better name here, fortunately it's still being marked as experimental.